Add rpc authentication support

facundovictor opened the pull request:
Add authentication support as is detailed in the [RPC API Contract](https://github.com/ivanalejandro0/cabinet/wiki/API-Contract).

Open vault method:
- **_URL:_** `http://{{CABINET_SERVER}}/api/<version>`
- **_HTTP Method:_** `post`
- **_Headers:_**

```
{
    'content-type' : 'application/json',
    'token' : <auth-token>
}
```
- **_Payload:_** 

```
Payload = {
    "jsonrpc": "2.0",
    "id": <request-number>,
    "method": "App.login",
    "params": {
        "username" : <user>,
        "password" : <pass>,
        "vault_path" : <vault_location>
    }
}
```

_**Response**_:

```
{
  "id": <request-number>,
  "jsonrpc": "2.0",
  "result": {
    success: true|false,
    error: undefined|{
      code: <error-code>,
      message: '<error-description>'
    }
  }
}
```

The `<token>` should be in the header of every request in order to be authenticated.

facundovictor committed:
Add support for token authentication

facundovictor committed:
Add login method

facundovictor committed:
Add token creator

Add a token generator based on firebase_token_generator

facundovictor committed:
Add salt generator

facundovictor committed:
Add secret key generator

facundovictor committed:
Remove JSONRPC's assumption

Flask JSONRPC makes an assumption on the arguments that should be
passed on every authenticated request. This forces the rpc client
to always send the username and password on every request.

facundovictor committed:
Move salt definition where it's used

facundovictor committed:
Move authenticate method to the rpc util

facundovictor committed:
Move the fixed JSONRPC class to the rpc package

ivanalejandro0 commented on rpc-server.py:
```diff
@@ -62,6 +88,7 @@ def get_random_open_port():
 
 
 if __name__ == '__main__':
+    app.secret_key = bcrypt.gensalt().decode('utf8')
```
Where do we use this `app.secret_key`?
Any reason te rely on `bcrypt` instead of using `os.urandom`?

facundovictor replied:
The [secret_key](http://flask.pocoo.org/docs/0.11/api/#flask.Flask.secret_key) is required for using [sessions](http://flask.pocoo.org/docs/0.11/quickstart/#sessions), used to sign cookies (where the session ID is stored).

About [bcrypt](https://github.com/pyca/bcrypt/blob/8825f7fb639a8beadeb9d91e290eabd6a9e1f02c/src/bcrypt/__init__.py#L42), you're right, for this case it should be `os.urandom`. I'll change this.

ivanalejandro0 commented:
The app token for the user who started the server must be displayed on stdout, otherwise there will be no way to use it.
You should either add the check for the API to use that token or (maybe better at this point) leave a `TODO` comment for add the check.

You are using a token for a session to keep the user data after login. I think that's an overkill since we already have an app token and only one user will use it at a time (the one who started the server). Is there a reason why we would like to have an extra verification layer?

ivanalejandro0 commented on rpc-server.py:
```diff
@@ -21,9 +24,31 @@
+
+@jsonrpc.method('App.login(username=str, password=str, vault_path=str) -> str')
+def login(username, password, vault_path):
+    generateToken(username, password, vault_path)
```
You're not doing any check here... you should add a comment about this accepting everything for now.

facundovictor replied:
Yes, indeed, my idea for this PR isn't to add the login check, but to add support for authenticating the requests that comes after login, using a token, instead of using the username and password every time.

ivanalejandro0 commented:
Would you mind adding a couple of calls to the `rpc-client.py` along this PR to show a call failing and working before and after a `login`?

facundovictor committed:
Change token generator

facundovictor committed:
Add util token_safeurl

Add the token_safeurl for generating new tokens. In the future
it could be replaced by secrets.token_safeurl.
Reference: https://docs.python.org/3/library/secrets.html

facundovictor committed:
Remove unused dependencies

facundovictor committed:
Modify all tokens to be generated by token_safeurl

facundovictor committed:
Edit rpc to publish the token and port to stdout

Edit the rpc server to publish once the generated token and random
port to the stdout. Thus, only the calling process will be able to
get it.

facundovictor committed:
Redirect stdout and stderr to log files

After publishing the token and random port, redirect the stdout
and stderr to the log files.

facundovictor committed:
Add password stored in session

facundovictor committed:
Add flush stdout after printing (token, port)

facundovictor committed:
Add example for starting the server

facundovictor committed:
Fix token and port recovering

facundovictor committed:
Change method login to open vault

facundovictor committed:
Add example call for open_vault

facundovictor commented:
The PR is updated:
- It uses only one token (for the user who started the server)
- The script `rpc-client.py` is updated as well, with an example of how to recover the token and port, and all the calls are updated to include the token in the header.
- The `login` method is changed to `open_vault`, and I've added a *TODO* message.

I'll be changing the decorator for authentication.

facundovictor committed:
Remove override of JSONRPC

facundovictor committed:
Add example for an unauthenticated call

facundovictor commented:
Hey @ivanalejandro0 , I've created a decorator for **authentication**, and removed the override of JSONRPC. Thus, we no longer have to worry about future changes related to authentication on the library flask-JSONRPC.

Also I've added an example (in `rpc-client.py`) of a call to the method `open_vault` with an authentication failure.
